docs/templates/index_template.md

```diff
@@ -0,0 +1,99 @@
+
+{INSERT_EXAMPLE}
+
+In this example we also optimize our hash by setting `preserves_structure=true`. You can do this any time your function is type stable, but there are additional conditions under which you can still set this flag to true, discussed below.
+
+[`StableHashTraits.Transformer`](@ref) takes a second positional argument which is the `StructType` you wish to use on the transformed return value. By default `StructType` is applied to the result to determine this automatically, but in some cases it can be useful to modify this trait by passing a second argument (see the example below).
+
+**To avoid StackOverflow errors** make sure you don't return the object itself as an element of some collection. It can be tempting to do e.g. `(mymetadata(x), x)` as a return value for `transformer`'s function. Instead you can use [`StableHashTraits.TransformIdentity`](@ref) to make sure this won't lead to an infinite regress: e.g. `(mymetadata(x), TransformIdentity(x))`. Using [`StableHashTraits.TransformIdentity`](@ref) will cause `x`'s transformed result to be `x` itself, thereby avoiding the infinite regress.
+
+`transformer` customizes how the *content* of your object is hashed. The hash of the type and any structure is customized separately. If you wish to customize how the type of an object is hashed, read on.
+
+## Using Contexts
+
+Because not every package knows about either `StableHashTraits` or `StructTypes`, there may be types you don't own that you want to customize the hash of. In this scenario you should define a context object that you pass as the second argument to `stable_hash` and define a method of `transformer` that dispatches on this context object as its second argument.
+
+A context is simply an arbitrary object that defines a method for
+[`StableHashTraits.parent_context`](@ref). By default the context to `stable_hash` is
+`HashVersion{version}()`. Because of `parent_context`, contexts can be stacked, and a
+`HashVersion` should be at the bottom of the stack. There are fallback methods for
+`transformer` that look at the value implemented by the parent context. In this way you need
+only define methods for the types you want to customize.
+
+For example, this customization makes the ordering of named tuple keys affect the hash value.
+
+```@doctest
+julia> begin
+        struct NamedTuplesOrdered{T}
+            parent::T
+        end
+        StableHashTraits.parent_context(x::NamedTuplesOrdered) = x.parent
+        function transformer(::Type{<:NamedTuple}, ::NamedTuplesOrdered)
+            Transformer(identity, StructTypes.OrderedStruct())
+        end
+        context = NamedTuplesOrdered(HashVersion{3}())
+       end;
+
+julia> stable_hash((; a=1:2, b=1:2), context) != stable_hash((; b=1:2, a=1:2), context)
+true
+```
+
+Without this context, the keys are first sorted because `StructType(NamedTuple) isa StructType.UnorderedStruct`.
```

Comment:
let's call this specific thing out in the README (named tuples are unordered for us) in a new section for "gotchas" or such. I think the semantics here are reasonable but it will help to highlight common gotchas.